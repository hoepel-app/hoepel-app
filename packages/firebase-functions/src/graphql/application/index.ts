import { resolvers } from './resolvers'
import { typeDef } from './typedef'

export const application = {
  resolvers,
  typeDef,
}
