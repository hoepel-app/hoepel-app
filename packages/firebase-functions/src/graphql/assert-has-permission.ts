import { NoPermissionError } from '../errors/no-permission.error'
import { Permission } from '@hoepel.app/types'
import * as admin from 'firebase-admin'

const db = admin.firestore()

export const assertHasPermission = async (
  uid: string,
  tenant: string,
  permission: Permission
): Promise<void> => {
  console.log(`Checking if ${uid} has permission ${permission}`)

  if (!uid) {
    throw new NoPermissionError('Uid not set')
  }

  try {
    const permissionsDoc = await db
      .collection('users')
      .doc(uid)
      .collection('tenants')
      .doc(tenant)
      .get()

    if (
      !permissionsDoc.exists ||
      !permissionsDoc?.data()?.permissions?.includes(permission)
    ) {
      throw new NoPermissionError(
        `No permission to access this resource (need ${permission} for tenant ${tenant})`,
        {
          permissionsDocExists: permissionsDoc.exists,
          permissionNeeded: permission,
          permissions: permissionsDoc.data()?.permissions,
          tenant,
        }
      )
    }
  } catch (err) {
    if (err instanceof NoPermissionError) {
      throw err // re-throw local error
    } else {
      throw new NoPermissionError(
        'Unexpected error checking permission',
        { tenant },
        err
      )
    }
  }
}
