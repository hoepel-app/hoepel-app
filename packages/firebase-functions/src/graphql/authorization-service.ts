import { Context, HoepelAppUser, ParentPlatformUser } from './index'

export class AuthorizationService {
  static assertLoggedInHoepelApp(
    context: Context
  ): asserts context is NonNullable<HoepelAppUser> {
    if (context == null || context.user == null || context.token == null) {
      throw new Error('Not logged in')
    }

    if (context.domain !== 'hoepel.app') {
      throw new Error('Logged in, but not a hoepel.app token')
    }
  }

  static assertLoggedInParentPlatform(
    context: Context
  ): asserts context is NonNullable<ParentPlatformUser> {
    if (context == null || context.user == null || context.token == null) {
      throw new Error('Not logged in')
    }

    if (context.domain !== 'parent-platform') {
      throw new Error('Logged in, but not a token for the parent platform')
    }
  }
}
