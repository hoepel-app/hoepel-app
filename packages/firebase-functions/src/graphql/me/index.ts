import { resolvers } from './resolvers'
import { typeDef } from './typedef'

export const me = {
  resolvers,
  typeDef,
}
