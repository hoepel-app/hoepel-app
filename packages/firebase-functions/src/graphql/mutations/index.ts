import { resolvers } from './resolvers'
import { typeDef } from './typedef'

export const mutations = {
  resolvers,
  typeDef,
}
