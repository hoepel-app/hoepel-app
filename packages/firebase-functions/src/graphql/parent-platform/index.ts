import { resolvers } from './resolvers'
import { typeDef } from './typedef'

export const parentPlatform = {
  resolvers,
  typeDef,
}
