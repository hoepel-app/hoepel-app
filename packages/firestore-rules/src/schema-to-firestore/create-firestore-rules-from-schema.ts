import { GraphQLSchema } from "graphql";
import { Rule } from "../firestore/rule";
import { getFirestoreCollectionsFromSchema } from "./get-firestore-collections-from-schema";

export const createFirestoreRulesFromSchema = (schema: GraphQLSchema, extraRules: readonly Rule[] = []): string => {
    const rulesFromSchemaAsText = getFirestoreCollectionsFromSchema(schema)
        .map(collection => collection.firestoreRules)
        .map(rules => rules.map(rule => rule.toString()).join('\n\n'))
        .join('\n\n');

    const extraRulesText = extraRules.map(rule => rule.toString()).join('\n\n')

    return [
        'service cloud.firestore {',
        indent('/// Extra rules provided for generation'),
        indent(extraRulesText),
        indent('/// Automaticaly generated Firestore rules'),
        indent(rulesFromSchemaAsText),
        '}'
    ].join('\n\n')
}

const indent = (str: string, indentation = '  '): string => str
    .split('\n')
    .map(line => indentation + line)
    .join('\n')
