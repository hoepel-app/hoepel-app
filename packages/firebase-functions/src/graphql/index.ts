import { me } from './me'
import { application } from './application'
import { mutations } from './mutations'
import { gql, ApolloServer } from 'apollo-server-express'
import * as admin from 'firebase-admin'
import { UserService } from '../services/user.service'
import { parseToken } from './parse-token'
import { IUser } from '@hoepel.app/types'
import * as Sentry from '@sentry/node'
import { tenant } from './tenant'

const db = admin.firestore()
const auth = admin.auth()

const userService = new UserService(db, auth)

const typeDef = gql`
  enum ReportType {
    ALL_CHILDREN
    ALL_CREW
    CHILDREN_WITH_COMMENT
    CREW_ATTENDANCES
    CHILD_ATTENDANCES
    FISCAL_CERTIFICATES_LIST
    CHILD_HEALTH_INSURANCE_CERTIFICATE
    CHILD_FISCAL_CERTIFICATE
    CHILD_INVOICE
    CHILDREN_PER_DAY
  }

  enum ReportFileFormat {
    XLSX
    PDF
    DOCX
  }

  type Report {
    id: ID!
    refPath: String!
    description: String!
    expires: String!
    created: String!
    format: ReportFileFormat!
    createdBy: String
    createdByUid: String!
    type: ReportType!
    childId: String
    year: Int
  }

  type Template {
    created: String!
    createdBy: String!
    fileName: String!
    displayName: String!
    type: ReportType!
  }

  type Query
  type Mutation
`

const resolvers = {
  ReportType: {
    ALL_CHILDREN: 'all-children',
    ALL_CREW: 'all-crew',
    CHILDREN_WITH_COMMENT: 'children-with-comment',
    CREW_ATTENDANCES: 'crew-attendances',
    CHILD_ATTENDANCES: 'child-attendances',
    FISCAL_CERTIFICATES_LIST: 'fiscal-certificates-list',
    CHILDREN_PER_DAY: 'children-per-day',
    CHILD_HEALTH_INSURANCE_CERTIFICATE: 'child-health-insurance-certificate',
    CHILD_FISCAL_CERTIFICATE: 'child-fiscal-certificate',
    CHILD_INVOICE: 'child-invoice',
  },
}

export type Context = {
  token?: admin.auth.DecodedIdToken
  user?: IUser
}

const getUserAndTokenFromHeader = async (
  authorizationHeader: string | null
): Promise<{ user: IUser; token: admin.auth.DecodedIdToken } | null> => {
  if (authorizationHeader == null) {
    return null
  }

  try {
    const decodedToken = await parseToken(authorizationHeader)
    const user = await userService.getUserFromDb(decodedToken.uid)

    if (user == null) {
      return null
    }

    return { token: decodedToken, user }
  } catch (err) {
    return null
  }
}

export const server = new ApolloServer({
  plugins: [
    {
      requestDidStart: () => {
        return {
          didEncounterErrors: async (requestContext) => {
            const header =
              requestContext.request.http?.headers.get('Authorization') ?? null

            const userInfo = await getUserAndTokenFromHeader(header)

            Sentry.configureScope((scope) => {
              if (userInfo != null) {
                scope.setUser({
                  email: userInfo.user.email,
                  username: userInfo.user.displayName,
                  id: userInfo.token.uid,
                })
              }

              requestContext.errors.forEach((err) => {
                console.error(err)
                Sentry.captureException(err)
              })
            })
          },
        }
      },
    },
  ],
  typeDefs: [
    typeDef,
    me.typeDef,
    mutations.typeDef,
    application.typeDef,
    tenant.typeDef,
  ],
  resolvers: [
    resolvers,
    me.resolvers,
    mutations.resolvers,
    application.resolvers,
    tenant.resolvers,
  ],
  introspection: true,
  playground: {
    endpoint: '/api/graphql',
  },
  tracing: true,
  context: async ({ req }): Promise<Context> => {
    return (
      (await getUserAndTokenFromHeader(req.headers.authorization ?? null)) || {}
    )
  },
})
