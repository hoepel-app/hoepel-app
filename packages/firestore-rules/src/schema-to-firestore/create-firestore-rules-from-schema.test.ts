import { buildSchema } from 'graphql';
import { createFirestoreRulesFromSchema } from './create-firestore-rules-from-schema';

const testSchema = buildSchema(`
enum COLLECTION_TYPE {
    INDEXED_BY_TENANT
    DOC_PER_TENANT
    DOC_PER_USER
}

directive @firestore(
    name: String!
    type: COLLECTION_TYPE!
    readPermission: String
    writePermission: String
    allowDelete: Boolean
    allowUpdate: Boolean
    allowPublicRead: Boolean
) on OBJECT

type AgeGroup @firestore(name:"age-groups", type:DOC_PER_TENANT, readPermission:"age-groups:read", writePermission:"age-groups:write") {
    id: ID!
    name: String
}

type Child @firestore(name:"children", type: INDEXED_BY_TENANT, readPermission:"child:read", writePermission:"child:write", allowUpdate: true, allowDelete: true) {
    id: ID!
    name: String
}

type Consumable @firestore(name:"consumables", type:DOC_PER_TENANT, readPermission:"consumables:read", writePermission:"consumables:write") {
    id: ID!
}

type Consumption @firestore(name:"consumptions", type:INDEXED_BY_TENANT, readPermission:"consumption:read", writePermission:"consumption:write", allowUpdate: true, allowDelete: true) {
    id: ID!
}

type User @firestore(name:"users", type:DOC_PER_USER) {
    id: ID!
}
`)

describe('createFirestoreRulesFromSchema', () => {
    it('create Firestore rules from schema', () => {
        expect(createFirestoreRulesFromSchema(testSchema)).toMatchSnapshot()
    })
})
