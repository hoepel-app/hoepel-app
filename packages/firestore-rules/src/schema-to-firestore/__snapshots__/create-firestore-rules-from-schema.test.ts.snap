// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFirestoreRulesFromSchema create Firestore rules from schema 1`] = `
"service cloud.firestore {

  /// Extra rules provided for generation

  

  /// Automaticaly generated Firestore rules

  // Generated permissions for collection 'age-groups' (collection contains docs with doc.id == tenant name)
  match /databases/{database}/documents/age-groups/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('age-groups:write', document);
    allow read  : if !exists(/databases/$(database)/documents/age-groups/$(document)) || hasPermission('age-groups:read', document);
    allow update: if hasPermission('age-groups:write', document);
    allow delete: if false;
  }
  
  // Generated permissions for collection 'children' (collection contains docs with doc.tenant == tenant name)
  match /databases/{database}/documents/children/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('child:write', request.resource.data.tenant);
    allow read  : if hasPermission('child:read', resource.data.tenant);
    allow update: if hasPermission('child:write', resource.data.tenant);
    allow delete: if hasPermission('child:write', resource.data.tenant);
  }
  
  // Generated permissions for collection 'consumables' (collection contains docs with doc.id == tenant name)
  match /databases/{database}/documents/consumables/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('consumables:write', document);
    allow read  : if !exists(/databases/$(database)/documents/consumables/$(document)) || hasPermission('consumables:read', document);
    allow update: if hasPermission('consumables:write', document);
    allow delete: if false;
  }
  
  // Generated permissions for collection 'consumptions' (collection contains docs with doc.tenant == tenant name)
  match /databases/{database}/documents/consumptions/{document} {
    function hasPermission(permission, tenant) {
      return permission in get(/databases/$(database)/documents/users/$(request.auth.uid)/tenants/$(tenant)).data.permissions;
    }
    allow create: if hasPermission('consumption:write', request.resource.data.tenant);
    allow read  : if hasPermission('consumption:read', resource.data.tenant);
    allow update: if hasPermission('consumption:write', resource.data.tenant);
    allow delete: if hasPermission('consumption:write', resource.data.tenant);
  }
  
  // Generated permissions for collection 'users' (collection contains docs with doc.id == user id)
  match /databases/{database}/documents/users/{uid} {
    allow create: if false;
    allow read  : if uid == request.auth.uid;
    allow update: if false;
    allow delete: if false;
  }
  
  // Generated permissions for collection 'users' (collection contains docs with doc.id == user id)
  match /databases/{database}/documents/users/{uid}/{document=**} {
    allow create: if false;
    allow read  : if uid == request.auth.uid;
    allow update: if false;
    allow delete: if false;
  }
  
  // Generated permissions for collection 'users' (collection contains docs with doc.id == user id) - Admin access
  match /databases/{database}/documents/users/{document=**} {
    allow create: if false;
    allow read  : if request.auth.token.isAdmin;
    allow update: if request.auth.token.isAdmin;
    allow delete: if false;
  }

}"
`;
